java-checks/src/main/java/org/sonar/java/checks/RedundantThrowsDeclarationCheck.java

```diff
@@ -216,14 +219,47 @@ private static boolean emptyBody(MethodTree methodTree) {
     if (block != null) {
       MethodInvocationVisitor visitor = new MethodInvocationVisitor();
       block.accept(visitor);
-      return visitor.thrownExceptions();
+      Set<Type> thrownExceptions = visitor.thrownExceptions();
+
+      if (thrownExceptions != null && methodTree.is(Tree.Kind.CONSTRUCTOR) && !visitor.visitedSuper) {
```

Comment:
I'm not sure we need this. II would prefer to let the logic of collecting exceptions hidden in the visitor. Why not directly handling this case in the `MethodInvocationVisitor` then?  You can perfectly give the symbol of the current methodTree as parameter of the `MethodInvocationVisitor` constructor, so you don't lose the info from where you are calling.

At the end of the visit, you could then simply add thrown exception from the default super constructor, if `super` has not been called and it's really a constructor.

Reply:
Yes, let's put this logic inside the visitor